### ISSUE TITLE
Fix handling of connections with holes >2gb

### ISSUE THREAD

**Issue by 0xxon** [zeek/zeek#1754]
When we have sequence number holes that are greater than INT_MAX, we currently loose track of this because our number range is not wide enough.

We should fix that, and also warn on excessively large holes in connections.

For details see #1709, which has a more thorough discussion.

**Comment by Stark-developer01** [comment]
Hi, can I contribute to fix this issue?

**Comment by 0xxon** [comment]
Hi,
this probably is not a great issue to start working on Zeek, since this is a quite central part of Zeek which might have quite subtle interactions with other things.

To find a good issue, perhaps search for the "good first issue" tag and grab one of them :)

https://github.com/zeek/zeek/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22

**Comment by 0xxon** [comment]
Ok, I just spent a whole bunch of time investigating this, and even made a branch that converted `net_util::seq_delta` to 64 bit integers, pulling all the functions that use it along.

After examining the resulting test failures, I realized that - from all that I can tell - the current behavior, using 32 bit unsigned integer mathematics, is completely correct, the behavior is desired as it currently is, and using 64 bit integers will make the code more difficult to read and harder to maintain.

To give a few more details: `seq_delta` substracts two uints from each other. The result of this operation is always well-defined, even in the case of a wrap-around.

The resulting value is converted into an int32_t. This produces the expected results when as long as the difference between the two numbers is < INT32_MAX. If the difference between the two sequence numbers is > INT32_MAX, you will get an, on a first glance, unexpected positive result after the substraction. However - this is also desired - the later code paths explicitly again check if a wrap-around happened and special-case this result. If we get rid of the 32-bit integers, we have to replicate this behavior - which is depended on to figure out wrap-arounds close to the end of the sequence-number-space.

So - in essence - the current behavior is correct and is used to determine sequence-number wrap arounds. If you have a connection where there are holes >2gb of size, then you might get unexpected results. But - thinking about this in detail and sketching around - when you have connections with holes >2gb, it gets undecidable if something is a "hole" or just a retransmit of earlier data. So - in my opinion, the behavior of Zeek is as correct as it can be in these cases.

Also note that there are only a few functions that purposefully use 32-bit numbers. The rest of the source-code that tracks total number of bytes in a connection has been using 64 bit numbers for a while.

Unless anyone chimes in here and lets me know that I am wrong, I am going to close this issue in ~a week.
